Compiler bug compiling an anonymous function

**Context**

I saw this in the context of an exercise that has to compute letter frequencies, and then sort them by some criteria that needs a function. This seems to be a minimal way to reproduce (please, ignore the warnings about unused variables).

**V doctor:**
```
OS: macos, macOS, 11.5.2, 20G95
Processor: 16 cpus, 64bit, little endian, Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
CC version: Apple clang version 12.0.5 (clang-1205.0.22.11)

getwd: /Users/fxn/AoC/2016/04
vmodules: /Users/fxn/.vmodules
vroot: /Users/fxn/v
vexe: /Users/fxn/v/v
vexe mtime: 2021-08-17 15:32:36
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.2 e1c762a

Git version: git version 2.32.0
Git vroot status: weekly.2021.33.2-11-ge1c762a6 (24 commit(s) behind V master)
.git/config present: true
thirdparty/tcc status: thirdparty-macos-amd64 689c8a02
```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg foo.v`
```v
mut frequencies := map[rune]int{}
sort_fn := fn [frequencies] (a &rune, b &rune) int {
    if frequencies[a] > frequencies[b] { return -1 }
    else if a > b { return -1 }
    else { return 1 }
}

```

**What did you expect to see?**

A successful compilation.


**What did you see instead?**
```
foo.v:2:1: warning: unused variable: `sort_fn`
    1 | mut frequencies := map[rune]int{}
    2 | sort_fn := fn [frequencies] (a &rune, b &rune) int {
      | ~~~~~~~
    3 |     if frequencies[a] > frequencies[b] { return -1 }
    4 |     else if a > b { return -1 }
==================
                                                                                                                                                                              *
/tmp/v/foo.14962202766298095660.tmp.c:10299:39: error: expected expression
        int (*sort_fn) (rune* a, rune* b) =     struct _V_anon_fn_cba870afa3c334d7_rune_rune__int_85_Ctx* _t1 = (struct _V_anon_fn_cba870afa3c334d7_rune_rune__int_85_Ctx*) memdup(&(struct _V_anon_fn_cba870afa3c334d7_rune_rune__int_85_Ctx){
                                                ^
/tmp/v/foo.14962202766298095660.tmp.c:10302:65: error: use of undeclared identifier '_t1'
__closure_create(anon_fn_cba870afa3c334d7_rune_rune__int_85, 3, _t1);
                                                                ^
2 warnings and 2 errors generated.
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

If you were not working with C interop, this is a compiler bug, please report the bug using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -425,20 +425,22 @@ fn (mut g Gen) gen_anon_fn(mut node ast.AnonFn) {
 		return
 	}
 	// it may be possible to optimize `memdup` out if the closure never leaves current scope
-	ctx_var := g.new_tmp_var()
+	// ctx_var := g.new_tmp_var()
 	cur_line := g.go_before_stmt(0)
 	ctx_struct := closure_ctx_struct(node.decl)
-	g.writeln('$ctx_struct* $ctx_var = ($ctx_struct*) memdup(&($ctx_struct){')
+
+	// TODO in case of an assignment, this should only call "__closure_set_data" and "__closure_set_function" (and free the former data)
+	g.write('__closure_create($node.decl.name, ${node.decl.params.len + 1}, ')
+	g.writeln('($ctx_struct*) memdup(&($ctx_struct){')
 	g.indent++
 	for var in node.inherited_vars {
 		g.writeln('.$var.name = $var.name,')
 	}
 	g.indent--
-	g.writeln('}, sizeof($ctx_struct));')
+	g.writeln('}, sizeof($ctx_struct)));')
 	g.empty_line = false
 	g.write(cur_line)
-	// TODO in case of an assignment, this should only call "__closure_set_data" and "__closure_set_function" (and free the former data)
-	g.write('__closure_create($node.decl.name, ${node.decl.params.len + 1}, $ctx_var)')
+	//g.write('__closure_create($node.decl.name, ${node.decl.params.len + 1}, $ctx_var)')
 }
 
 fn (mut g Gen) gen_anon_fn_decl(mut node ast.AnonFn) {

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -426,7 +426,7 @@ fn (mut g Gen) gen_anon_fn(mut node ast.AnonFn) {
 	}
 	// it may be possible to optimize `memdup` out if the closure never leaves current scope
 	// ctx_var := g.new_tmp_var()
-	cur_line := g.go_before_stmt(0)
+	// cur_line := g.go_before_stmt(0)
 	ctx_struct := closure_ctx_struct(node.decl)
 
 	// TODO in case of an assignment, this should only call "__closure_set_data" and "__closure_set_function" (and free the former data)
@@ -437,9 +437,9 @@ fn (mut g Gen) gen_anon_fn(mut node ast.AnonFn) {
 		g.writeln('.$var.name = $var.name,')
 	}
 	g.indent--
-	g.writeln('}, sizeof($ctx_struct)));')
+	g.write('}, sizeof($ctx_struct)))')
 	g.empty_line = false
-	g.write(cur_line)
+	// g.write(cur_line)
 	//g.write('__closure_create($node.decl.name, ${node.decl.params.len + 1}, $ctx_var)')
 }
 

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -425,10 +425,7 @@ fn (mut g Gen) gen_anon_fn(mut node ast.AnonFn) {
 		return
 	}
 	// it may be possible to optimize `memdup` out if the closure never leaves current scope
-	// ctx_var := g.new_tmp_var()
-	// cur_line := g.go_before_stmt(0)
 	ctx_struct := closure_ctx_struct(node.decl)
-
 	// TODO in case of an assignment, this should only call "__closure_set_data" and "__closure_set_function" (and free the former data)
 	g.write('__closure_create($node.decl.name, ${node.decl.params.len + 1}, ')
 	g.writeln('($ctx_struct*) memdup(&($ctx_struct){')
@@ -439,8 +436,6 @@ fn (mut g Gen) gen_anon_fn(mut node ast.AnonFn) {
 	g.indent--
 	g.write('}, sizeof($ctx_struct)))')
 	g.empty_line = false
-	// g.write(cur_line)
-	//g.write('__closure_create($node.decl.name, ${node.decl.params.len + 1}, $ctx_var)')
 }
 
 fn (mut g Gen) gen_anon_fn_decl(mut node ast.AnonFn) {

--- a/vlib/v/tests/closure_test.v
+++ b/vlib/v/tests/closure_test.v
@@ -155,11 +155,14 @@ fn test_go_call_closure() {
 fn test_closures_with_ifstmt() {
 	a := 1
 	f := fn [a] (x int) int {
-		if a > x { return 1 }
-		else { return -1 }
+		if a > x { return 1
+		 } else { return -1
+		 }
 	}
 	g := fn [a] () int {
-		if true { return 1 }
+		if true {
+			return 1
+		}
 		return 0
 	}
 	assert f(0) == 1

--- a/vlib/v/tests/closure_test.v
+++ b/vlib/v/tests/closure_test.v
@@ -151,3 +151,16 @@ fn test_go_call_closure() {
 	assert <-ch == 15
 }
 */
+
+fn test_closures_with_ifstmt() {
+	a := 1
+	f := fn [a] (x int) int {
+		if a > x { return 1 }
+		else { return -1 }
+	}
+	g := fn [a] () int {
+		return 1
+	}
+	assert f(0) == 1
+	assert g() == 1
+}

--- a/vlib/v/tests/closure_test.v
+++ b/vlib/v/tests/closure_test.v
@@ -159,7 +159,8 @@ fn test_closures_with_ifstmt() {
 		else { return -1 }
 	}
 	g := fn [a] () int {
-		return 1
+		if true { return 1 }
+		return 0
 	}
 	assert f(0) == 1
 	assert g() == 1

